ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PipeChainTypeHintStage.fs

```diff
@@ -24,22 +28,54 @@ type SameLinePipeHints =
     | Hide
 
 type PipeOperatorVisitor(sameLinePipeHints: SameLinePipeHints) =
-    inherit TreeNodeVisitor<List<IReferenceExpr * ITreeNode>>()
+    inherit TreeNodeVisitor<List<IReferenceExpr * ITreeNode * bool>>()
 
     let showSameLineHints =
         match sameLinePipeHints with
         | SameLinePipeHints.Show -> true
         | SameLinePipeHints.Hide -> false
 
-    let visitBinaryAppExpr binaryAppExpr (context: List<IReferenceExpr * ITreeNode>) =
-        if not (isPredefinedInfixOpApp "|>" binaryAppExpr) then () else
+    let isApplicable binaryAppExpr =
+        isPredefinedInfixOpApp "|>" binaryAppExpr
 
-        let exprToAdorn = binaryAppExpr.LeftArgument
-        if isNull exprToAdorn then () else
+    let visitBinaryAppExpr binaryAppExpr (context: List<IReferenceExpr * ITreeNode * bool>) =
+        if not (isApplicable binaryAppExpr) then () else
 
         let opExpr = binaryAppExpr.Operator
-        if showSameLineHints || exprToAdorn.EndLine <> opExpr.StartLine then
-            context.Add(opExpr, exprToAdorn :> _)
+
+        let leftArgument = binaryAppExpr.LeftArgument
+        let rightArgument = binaryAppExpr.RightArgument
+
+        if isNull leftArgument then () else
+
+        if showSameLineHints || leftArgument.EndLine <> opExpr.StartLine then
+            context.Add(opExpr, leftArgument, true)
+
+        if isNull rightArgument || rightArgument :? IFromErrorExpr then () else
+
+        let isTopBinary =
+            let outermostParent =
+                binaryAppExpr
+                 .IgnoreParentParens()
+                 .GetOutermostParentExpressionFromItsReturn()
+                 .IgnoreParentParens()
```

Comment:
I wonder if this parens skipping would be better to be inside `GetOutermostParentExpressionFromItsReturn` instead of having to do it twice later?